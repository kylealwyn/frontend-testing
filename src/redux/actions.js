import fetch from 'isomorphic-fetch';
import * as Types from './types';

export const addTodo = (text) => ({
  type: Types.ADD_TODO,
  payload: text,
});

function fetchTodosRequest() {
  return {
    type: Types.FETCH_TODOS_REQUEST
  }
}

function fetchTodosSuccess(body) {
  return {
    type: Types.FETCH_TODOS_SUCCESS,
    payload: body
  }
}

function fetchTodosFailure(error) {
  return {
    type: Types.FETCH_TODOS_FAILURE,
    payload: error
  }
}

export function fetchTodos() {
  return dispatch => {
    dispatch(fetchTodosRequest())
    return fetch('http://example.com/todos')
      .then(res => res.json())
      .then(json => dispatch(fetchTodosSuccess(json.body)))
      .catch(ex => dispatch(fetchTodosFailure(ex)))
  }
}
