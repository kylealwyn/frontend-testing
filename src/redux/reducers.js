import { ADD_TODO } from './types'

export const initialTodoState = []

export default function todos(state = initialTodoState, action) {
  switch (action.type) {
    case ADD_TODO:
      return [
        {
          id: state.reduce((maxId, todo) => Math.max(todo.id, maxId), -1) + 1,
          completed: false,
          text: action.text
        },
        ...state
      ]

    default:
      return state
  }
}
