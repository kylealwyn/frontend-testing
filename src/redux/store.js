import { createStore, applyMiddleware } from 'redux';
import thunkMiddleware from 'redux-thunk';
import todoReducer from './reducers';

export const configureStore = () => {
  const store = createStore(todoReducer, {}, applyMiddleware(
   thunkMiddleware,
 ))

 return store;
}
