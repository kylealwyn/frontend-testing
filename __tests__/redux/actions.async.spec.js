import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import * as actions from '../../src/redux/actions'
import * as types from '../../src/redux/types'
import nock from 'nock'

describe('async actions', () => {
  const middlewares = [ thunk ]
  const mockStore = configureMockStore(middlewares)
  let store;

  beforeEach(() => {
    store = mockStore({ todos: [] });

    nock('http://example.com/')
      .get('/todos')
      .reply(200, { body: { todos: ['do something'] }})
  });

  afterEach(() => {
    nock.cleanAll()
  })

  it('creates FETCH_TODOS_SUCCESS when fetching todos has been done', () => {
    return store.dispatch(actions.fetchTodos())
      .then(() => { // return of async actions
        expect(store.getActions()).toMatchSnapshot()
      })
  })
})
