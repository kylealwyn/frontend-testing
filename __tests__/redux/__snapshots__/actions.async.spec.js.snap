// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`async actions creates FETCH_TODOS_SUCCESS when fetching todos has been done 1`] = `
Array [
  Object {
    "type": "FETCH_TODOS_REQUEST",
  },
  Object {
    "payload": Object {
      "todos": Array [
        "do something",
      ],
    },
    "type": "FETCH_TODOS_SUCCESS",
  },
]
`;
